Issue #171: Show placeholder text

aforcier opened the pull request:
Fixes #171. Displays placeholder text for empty fields.

![171-placeholder-visual](https://cloud.githubusercontent.com/assets/9613966/8756418/1f93955e-2ca0-11e5-9271-945f584b00b8.png)

![171-placeholder-html](https://cloud.githubusercontent.com/assets/9613966/8756417/1bf5c2a0-2ca0-11e5-92cd-42d7f0d14924.png)

(no placeholder text for content in HTML mode, reflecting iOS)

Had some trouble with autocorrect (API 19+), the placeholder color would remain and color the actual text entered until space or entered was pressed. Fixed by setting a fixed color in the CSS and dropping the `ZSSEditor` methods that were handling setting it dynamically.

Note that the text in the editor is coming from the example app - the actual placeholders will be set from WPAndroid, so we can have 'Post Title' or 'Page Title' accordingly, as with iOS.

aforcier committed:
Added support for setting the title and content placeholder text

aforcier committed:
Added placeholder text loading to the example activity

aforcier committed:
Added coloring for placeholder text in HTML mode

aforcier committed:
Updated editor CSS with placeholder color

- Fixes an issue on API19+ (autocorrect) where the placeholder color would remain after text was typed in.
- With placeholder color no longer set dynamically, dropped functions that were no longer needed

aforcier committed:
Fixed a broken test (missing abstract method implementations)

aforcier committed:
Merge branch 'develop' into issue/171-placeholder-text

# Conflicts:
#	WordPressEditor/src/main/java/org/wordpress/android/editor/EditorFragment.java
#	WordPressEditor/src/main/res/layout/fragment_editor.xml

aforcier committed:
Updated tablet layout xml with placeholder changes

aforcier commented:
cc @bummytime, @thianhlu

bummytime commented:
@aforcier In my testing so for, API 22 looks good, however API 16 (simulator) and 17 (simulator and device) had a few issues:

https://cloudup.com/clwCdpvn7mR
- The initial cursor position seems to be shifted left and up.
- The select all and cut behavior seems a little off. On new posts it seems to work fine, but on our existing test post the placeholder never seems to come back unless backspace is pressed multiple times.

bummytime commented on libs/editor-common/assets/editor.css:
```diff
@@ -287,6 +287,7 @@ div.field[contenteditable] {
 
 div.field[placeholderText][contenteditable=true]:empty:before {
     content: attr(placeholderText);
+    color: #87a6bc;
```
A lot easier to do it here, eh? :-)

aforcier replied:
One picks one's battles with Android autocorrect.

bummytime commented on WordPressEditor/src/main/res/values/colors.xml:
```diff
@@ -10,5 +10,7 @@
     <color name="format_bar_button_normal_color">@color/wp_gray</color>
     <color name="format_bar_button_highlighted_color">@color/wp_blue</color>
 
+    <color name="placeholder_text">@color/wp_gray</color>
```
Just thinking out loud...do you think we need to be more specific with the name here since this value defines the placeholder color in the source view and not the visual view?

aforcier replied:
Agreed, I'll rename that.

aforcier commented:
> The initial cursor position seems to be shifted left and up.

Nice catch, looks like it's indeed caused by setting a placeholder. Investigating.

> The select all and cut behavior seems a little off. On new posts it seems to work fine, but on our existing test post the placeholder never seems to come back unless backspace is pressed multiple times.

I'm able to reproduce this on an API 22 device. From inspection, it looks like Select All > Cut leaves behind some tags (usually a set of p tags, sometimes a `<br/>`), and so the field isn't understood to be empty by the CSS and the placeholder doesn't show. The wrapping tags show up when loading existing content and also when switching back and forth with HTML mode. Looking into a fix.

aforcier committed:
Fixed a bug where the placeholder text would not re-appear after deleting all the contents of a field in visual mode

- Selecting and deleting all would leave in a carriage return character, which is now looked for and cleared
- Also changed to using setPlainText() instead of setHTML() for the title

aforcier committed:
Renamed the HTML mode placeholder text color resource for clarity

aforcier committed:
Re-assigned content field padding to body and separator in editor.css

- Fixes a bug where the cursor would ignore field padding for empty fields with placeholders on API<19

aforcier commented:
Updates:
- Fixed the cursor position bug.

It seems to be related to #179, as another oddity of empty fields for `API<19`. The contenteditable's padding CSS is ignored when there is placeholder text and the document falls back to the body CSS. I fixed this by moving the contenteditable padding to the body and the separator div. This solves the issue and doesn't affect the layout at all. However, focus is still lost if you delete the last character in the field, and you have to re-focus; this has the same cause as #179 and will be fixed at a later time.
- Fixed the Select All > Cut bug

Deleting all the contents of the document was sometimes leaving behind a carriage return character. The editor now checks if the field is empty except for a carriage return and empties it, re-displaying the placeholder text.
- Rename placeholder text color resource

Ready for another look @bummytime!

bummytime commented:
Ok @aforcier. Nice updates. :shipit:
